File: Fall 2024/ma553/homework/hw1/heat2d.f90
!
! diff2d.f90 - Simulates two-dimensional diffusion on a square domain
!              Fortran 90 version
!
!
program diffusion_2d
    implicit none
    ! Include file sets parameters D, x1, x2, runtime, dx, outtime, and graphics:
    logical, parameter :: False=.false., True=.true.
    real*8  :: D, x1, x2, runtime, dx, outtime, dt, simtime, a, b
    logical :: graphics
    integer :: Nx, Ny, npnts, nsteps, nper, i, j, s
    real*8, allocatable, dimension(:)             :: x
    real*8, allocatable, dimension(:,:)           :: laplacian
    real*8, pointer, dimension(:,:)               :: u, u_next, tmp
    real*8, target, allocatable, dimension(:,:,:) :: u_buffer
    real*8  :: tstart, tend
    include 'diff2dparams.py'

    ! Compute derived parameters.
    Nx     = floor((x2-x1)/dx) ! number of x points
    Ny     = Nx           ! number of y points, same as x in this case.
    npnts  = Nx + 2       ! number of x points including boundary points
    dx     = (x2-x1)/Nx   ! recomputed (previous dx may not fit in [x1,x2])
    dt     = 0.25*dx**2/D    ! time step size (edge of stability)
    nsteps = int(runtime/dt) ! number of steps of that size to reach runtime
    nper   = int(outtime/dt) ! how many step s between snapshots
    if (nper == 0) then
       nper = 1
    endif
    ! Allocate arrays.
    allocate(x(npnts))
    allocate(u_buffer(npnts,npnts,2))
    allocate(laplacian(npnts,npnts))
    u => u_buffer(:,:,1)
    u_next => u_buffer(:,:,2)
    ! Initialize.
    do i=1,npnts 
       x(i) = x1+((i-2)*(x2-x1))/Nx ! x values (also y values)
    enddo
    u=0.0
    u_next=0.0
    simtime=0*dt
    do i=2,npnts-1
        a = 1 - abs(1 - 4*abs((x(i)-(x1+x2)/2)/(x2-x1)))
        do j=2,npnts-1
            b = 1 - abs(1 - 4*abs((x(j)-(x1+x2)/2)/(x2-x1)))
            u(i,j) = a*b
        end do
    end do
    ! Output initial signal.
    print '(" at t =  ", f5.2 )', simtime    
    if (graphics .eqv. .true.) then
       do i=1,npnts
          do j=1,npnts
             write(7,*) x(i), x(j), u(i,j)
          enddo
       enddo
    endif
    call cpu_time(tstart)
    ! Let temporary array hold laplacian.
    laplacian = 0.0 
    do s=1,nsteps
        ! Take one time step to produce new density.
        ! Using explicit iterations over indices instead
        do j=2,Ny+1
            do i=2,Nx+1
                laplacian(i,j) = u(i+1,j)+u(i-1,j)+u(i,j+1)+u(i,j-1)-4*u(i,j)
            end do
        end do
        do j=2,Ny+1
            do i=2,Nx+1
                u_next(i,j) = u(i,j) + (D/dx**2)*dt*laplacian(i,j)
            end do
        end do
        ! Swap array pointers so t+1 becomes the new t, and update simulation time.
        tmp => u
        u => u_next
        u_next => tmp
        simtime = simtime + dt
        ! Plot and report time.
        if (mod(s,nper) == 0) then
            print '(" at t =  ", f5.2 )', simtime    
            if (graphics .eqv. .true.) then
               do i=1,npnts
                  do j=1,npnts
                     write(11,*) x(i), x(j), u(i,j)
                  enddo
               enddo
               write(11,*)
        !        call plotdens(dens, x(1), x(npnts), .false.)
            endif
        endif
    enddo
    if (graphics .eqv. .true.) then
       do i=1,npnts
          do j=1,npnts
             write(11,*) x(i), x(j), u(i,j)
          enddo
       enddo
    endif
    call cpu_time(tend)
    print '("Simulation finished in ", f16.12," s")', tend-tstart    
end program diffusion_2d

File: Fall 2024/ma553/homework/hw1/heat2d.py
#!/usr/bin/env python
#
# heat2d.py - Simulates two-dimensional heat equation on a square domain
#             This one is pure python, it does not use numpy.
#

import time

from heat2dparams import D, Nx, Ny, plot, tmax, tout, xmax, xmin, ymax, ymin
from heat2dplot import plotsol


def initialize(u, x, y):
    """Apply the initial condition u(x,y,0)=u0(x,y)"""
    nx = len(x)
    ny = len(y)
    for i in range(1, nx - 1):
        ui = 1 - abs(1 - 4 * abs((x[i] - (x[0] + x[-1]) / 2) / (x[-1] - x[0])))
        for j in range(1, ny - 1):
            uj = 1 - abs(1 - 4 * abs((y[j] - (y[0] + y[-1]) / 2) / (y[-1] - y[0])))
            u[i][j] = ui * uj

def evolve(u_new, u, dt, dx2, dy2):
    """Update PDE by advancing one time step with size dt."""
    nx = len(u)
    ny = len(u[0])
    for i in range(1, nx - 1):
        for j in range(1, ny - 1):
            laplacian = (u[i + 1][j] - 2 * u[i][j] + u[i - 1][j]) / dx2 + (
                u[i][j + 1] - 2 * u[i][j] + u[i][j - 1]
            ) / dy2
            u_new[i][j] = u[i][j] + D * dt * laplacian

def iterate(u_new, u, dt, dx2, dy2, nsteps, plot_interval):
    """Run fixed number of time steps of heat equation"""
    simtime = 0 * dt
    print(' at t = {0:5.2f}'.format(simtime))
    if plot:
        plotsol(u, xmin, xmax, ymin, ymax, first=True)
    for s in range(nsteps + 1):
        evolve(u_new, u, dt, dx2, dy2)  # Advance one step in time
        # Swap array pointers so t+1 becomes the new t, and update simulation time.
        u, u_new = u_new, u
        simtime += dt
        # Plot and report time.
        if (s + 1) % plot_interval == 0:
            print(' at t = {0:5.2f}'.format(simtime))
            if plot:
                plotsol(u, xmin, xmax, ymin, ymax)

# the driver routine
def main():
    # Parameters set in heat2dparams.py:
    dx = (xmax - xmin) / Nx
    dy = (ymax - ymin) / Ny
    dx2 = dx**2
    dy2 = dy**2
    dt = (dx2 * dy2) / (2 * D * (dx2 + dy2))  # time step size (edge of stability)
    nsteps = int(tmax / dt)  # number of steps of that size to reach tmax
    plot_interval = int(tout / dt)  # how many steps between snapshots
    if plot_interval == 0:
        plot_interval = 1

    # Allocate arrays x, y, u, u_new.
    x = [xmin + i * dx for i in range(Nx + 1)]  # x values
    y = [ymin + i * dy for i in range(Ny + 1)]  # y values
    u = [[0.0] * (Ny + 1) for i in range(Nx + 1)]  # time step t
    u_new = [[0.0] * (Ny + 1) for i in range(Nx + 1)]  # time step t+1

    # Initialize and output/plot initial profile.
    initialize(u, x, y)
    print("\nHeat equation solver")
    print("Parameters")
    print("----------")
    print("  D={}, xmin={}, xmax={}, ymin={}, ymax={}".format(D, xmin, xmax, ymin, ymax))
    print("  dx={}, dy={}, dt={}".format(dx, dy, dt))
    print("  total time steps={}, plot interval={}\n".format(nsteps, plot_interval))

    # Iterate and plot solution
    t0 = time.time()
    iterate(u_new, u, dt, dx2, dy2, nsteps, plot_interval)
    t1 = time.time()

    print('\tat t = {0:5.2f}'.format(tmax))
    print("\nSimulation finished in {0} s".format(t1 - t0))


if __name__ == '__main__':
    main()
